chapter3/challenges/program/lib/password_manager_2.rb

```diff
@@ -1,2 +1,54 @@
 class PasswordManager2
-end
\ No newline at end of file
+    def initialize
+        #Initializes three arrays, password manager being the main one, and two other arrays for checking / ease of ordering.
+        @password_manager = Array.new()
+        @password_array = Array.new()
+        @service_array = Array.new()
+    end
+        def add(service, password)
+           new_entry = {
+               "service" => service,
+               "password" => password,
+               "added" => DateTime.now()
+           }
+   
+           if @service_array.include? service
+               return "ERROR: Service names must be unique"
+           elsif @password_array.include? password 
+               return "ERROR: Passwords must be unique"
+           else 
+               @password_manager << new_entry
+               @service_array << service
+               @password_array << password
+           end 
+        end 
+       
+       def services
+           return @service_array
+       end
+       
+       def remove(service)
+        @service_array.reject!{ |e| e[service] == service }
```

Comment:
I like the .reject method, since it more clearly describes what is going on. I may pinch that, to replace my .filter! 
But your code only removes the service from the list of services. It would still be possible to get access to the saved password from manager. It passes the tests, so it meets the requirements in that sense. But it isn't a complete deletion of the service. A service could end up listed in the manager twice now, (add, remove then add) and when you search for the password you will get the first result -- which has an old password.

Reply:
Yup just realised that and think applied a fix.